src/Program.ts

```diff
@@ -307,15 +348,23 @@ export class Program {
 
                 //add the file to the program
                 this.files[pathAbsolute] = brsFile;
-                let fileContents = await getFileContents();
+                let fileContents = {
+                    pathAbsolute: pathAbsolute,
+                    src: await getFileContents()
+                };
+                this.emit('file-source-loaded', fileContents);
+
                 this.logger.time(LogLevel.info, ['parse', chalk.green(pathAbsolute)], () => {
-                    brsFile.parse(fileContents);
+                    brsFile.parse(
+                        fileContents.src,
+                        () => this.emit('file-parsed', brsFile)
+                    );
                 });
                 file = brsFile;
 
                 this.dependencyGraph.addOrReplace(brsFile.dependencyGraphKey, brsFile.ownScriptImports.map(x => x.pkgPath.toLowerCase()));
 
-
+                this.emit('file-validated', brsFile);
```

Comment:
That's exactly the code I had initially. As you see, such a generic function is tricky to type properly - the `any[]` is quite painful to see ;)

Reply:
To avoid using `any[]` we can overload the function definition to deal with that. 

```
private emitPluginEvent(eventName: 'first-event-type', data: string);
private emitPluginEvent(eventName: 'second-event-type', data: {srcDestEntry: SourceDestEntry, file: File});
private emitPluginEvent(eventName: 'third-event-type', data: boolean);
private emitPluginEvent(eventName: string, data: any[]) {
```

Reply:
What do you think of the API now though? You can listen to all the events directly on the ProgramBuilder.